core/src/layers/prometheus.rs

```diff
@@ -0,0 +1,990 @@
+            inner,
+            stats: Arc::new(PrometheusMetrics::new(registry.clone())),
+        }
+    }
+}
+/// [`PrometheusMetrics`] provide the performance and IO metrics.
+#[derive(Debug)]
+pub struct PrometheusMetrics {
+    // metadata
+    pub metadata_request_counts: GenericCounter<AtomicU64>,
+    pub metadata_request_latency: Histogram,
+
+    // create
+    pub create_request_counts: GenericCounter<AtomicU64>,
+    pub create_request_latency: Histogram,
+
+    /// read
+    pub read_request_counts: GenericCounter<AtomicU64>,
+    pub read_request_latency: Histogram,
+    pub read_size: Histogram,
+
+    // write
+    pub write_request_counts: GenericCounter<AtomicU64>,
+    pub write_request_latency: Histogram,
+    pub write_size: Histogram,
+
+    // stat
+    pub stat_request_counts: GenericCounter<AtomicU64>,
+    pub stat_request_latency: Histogram,
+
+    // delete
+    pub delete_request_counts: GenericCounter<AtomicU64>,
+    pub delete_request_latency: Histogram,
+
+    // list
+    pub list_request_counts: GenericCounter<AtomicU64>,
+    pub list_request_latency: Histogram,
+
+    // scan
+    pub scan_request_counts: GenericCounter<AtomicU64>,
+    pub scan_request_latency: Histogram,
+
+    // presign
+    pub presign_request_counts: GenericCounter<AtomicU64>,
+    pub presign_request_latency: Histogram,
+
+    // batch
+    pub batch_request_counts: GenericCounter<AtomicU64>,
+    pub batch_request_latency: Histogram,
+
+    // blocking create
+    pub blocking_create_request_counts: GenericCounter<AtomicU64>,
+    pub blocking_create_request_latency: Histogram,
+
+    // blocking read
+    pub blocking_read_request_counts: GenericCounter<AtomicU64>,
+    pub blocking_read_request_latency: Histogram,
+    pub blocking_read_size: Histogram,
+
+    // blocking write
+    pub blocking_write_request_counts: GenericCounter<AtomicU64>,
+    pub blocking_write_request_latency: Histogram,
+    pub blocking_write_size: Histogram,
+
+    // blocking stat
+    pub blocking_stat_request_counts: GenericCounter<AtomicU64>,
+    pub blocking_stat_request_latency: Histogram,
+
+    // blocking delete
+    pub blocking_delete_request_counts: GenericCounter<AtomicU64>,
+    pub blocking_delete_request_latency: Histogram,
+
+    // blocking list
+    pub blocking_list_request_counts: GenericCounter<AtomicU64>,
+    pub blocking_list_request_latency: Histogram,
+
+    // blocking scan
+    pub blocking_scan_request_counts: GenericCounter<AtomicU64>,
+    pub blocking_scan_request_latency: Histogram,
+}
+
+impl PrometheusMetrics {
+    /// new with prometheus register.
+    pub fn new(registry: Registry) -> Self {
+        // metadata
+        let metadata_request_counts = register_int_counter_with_registry!(
+            "metadata_request_counts",
+            "Total times of metadata be called",
+            registry
+        )
+        .unwrap();
+        let opts = histogram_opts!(
+            "metadata_request_latency",
```

Comment:
I prefer to following the naming style as described at https://prometheus.io/docs/practices/naming/

- opendal_requests_total
- opendal_requests_duration_seconds
- opendal_bytes_total